File: plinko-shared/src/protocol.rs
use crate::data::constraint::DateConstraint;
use crate::data::dependency::Dependency;
use crate::data::ids::{MilestoneId, NodeId, TagId, TaskId, UserId};
use crate::data::plan::DependencyError;
use crate::data::scheduler::SchedulerError;
use crate::data::task::WorkerSlot;
use crate::data::{Milestone, Plan, Status, Task, User, WorkSchedule};
use crate::monday::{BoardColumn, MondayConfig, MondayUser};
use chrono::NaiveDate;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use uuid::Uuid;

// ── Auth shared types ────────────────────────────────────────────────────────

/// A login user record (no password hash).
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct AuthUser {
    pub id: String,
    pub email: String,
    pub is_admin: bool,
}

/// Role within an organisation.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub enum OrgRole {
    Admin,
    User,
    Viewer,
}

/// A user's membership in an organisation.
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct OrgMembership {
    pub org_id: String,
    pub org_name: String,
    pub role: OrgRole,
}

/// Summary of an organisation (for lists).
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Organisation {
    pub id: String,
    pub name: String,
}

/// A member of an organisation with their role.
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct OrgMember {
    pub user_id: String,
    pub email: String,
    pub role: OrgRole,
}

/// Per-plan permission entry for a specific user (used in the plan-access management UI).
/// `permission` is one of `"NoAccess"`, `"Viewer"`, `"User"`, or `"Default"` (inherit org role).
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct PlanPermissionEntry {
    pub plan_id: uuid::Uuid,
    pub plan_name: String,
    pub permission: String,
}

/// Maps a login user UUID to a plan user UUID.
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct UserLink {
    pub login_user_id: Uuid,
    pub plan_user_id: UserId,
}

/// Deserializes an `Option<Option<T>>` field so that an absent JSON field
/// produces `None` (no-op / don't update) while an explicit JSON `null`
/// produces `Some(None)` (clear the value).  Without this, both cases
/// round-trip as `None` because serde maps JSON `null` to the outer `None`.
fn deserialize_optional_field<'de, T, D>(de: D) -> Result<Option<Option<T>>, D::Error>
where
    D: serde::Deserializer<'de>,
    T: serde::Deserialize<'de>,
{
    Ok(Some(Option::deserialize(de)?))
}

#[derive(Default, Serialize, Deserialize, Clone, Debug)]
pub struct TaskPatch {
    pub name: Option<String>,
    pub description: Option<String>,
    pub status: Option<Status>,
    #[serde(default, deserialize_with = "deserialize_optional_field")]
    pub actual_start_date: Option<Option<NaiveDate>>,
    #[serde(default, deserialize_with = "deserialize_optional_field")]
    pub actual_end_date: Option<Option<NaiveDate>>,
    #[serde(default, deserialize_with = "deserialize_optional_field")]
    pub constraint: Option<Option<DateConstraint>>,
    pub duration_days_target: Option<f32>,
    pub workers: Option<Vec<WorkerSlot>>,
    pub dependencies: Option<Vec<Dependency>>,
    pub relaxed_mode: Option<bool>,
}

impl TaskPatch {
    pub fn new() -> Self {
        Self::default()
    }
    pub fn name(mut self, v: impl Into<String>) -> Self {
        self.name = Some(v.into());
        self
    }
    pub fn description(mut self, v: impl Into<String>) -> Self {
        self.description = Some(v.into());
        self
    }
    pub fn status(mut self, v: Status) -> Self {
        self.status = Some(v);
        self
    }
    pub fn actual_start_date(mut self, v: Option<NaiveDate>) -> Self {
        self.actual_start_date = Some(v);
        self
    }
    pub fn actual_end_date(mut self, v: Option<NaiveDate>) -> Self {
        self.actual_end_date = Some(v);
        self
    }
    pub fn constraint(mut self, v: Option<DateConstraint>) -> Self {
        self.constraint = Some(v);
        self
    }
    pub fn duration_days_target(mut self, v: f32) -> Self {
        self.duration_days_target = Some(v);
        self
    }
    pub fn workers(mut self, v: Vec<WorkerSlot>) -> Self {
        self.workers = Some(v);
        self
    }
    pub fn dependencies(mut self, v: Vec<Dependency>) -> Self {
        self.dependencies = Some(v);
        self
    }
    pub fn relaxed_mode(mut self, v: bool) -> Self {
        self.relaxed_mode = Some(v);
        self
    }
}

#[derive(Default, Serialize, Deserialize, Clone, Debug)]
pub struct MilestonePatch {
    pub name: Option<String>,
    pub description: Option<String>,
    #[serde(default, deserialize_with = "deserialize_optional_field")]
    pub constraint: Option<Option<DateConstraint>>,
    pub dependencies: Option<Vec<Dependency>>,
}

impl MilestonePatch {
    pub fn new() -> Self {
        Self::default()
    }
    pub fn name(mut self, v: impl Into<String>) -> Self {
        self.name = Some(v.into());
        self
    }
    pub fn description(mut self, v: impl Into<String>) -> Self {
        self.description = Some(v.into());
        self
    }
    pub fn constraint(mut self, v: Option<DateConstraint>) -> Self {
        self.constraint = Some(v);
        self
    }
    pub fn dependencies(mut self, v: Vec<Dependency>) -> Self {
        self.dependencies = Some(v);
        self
    }
}

#[derive(Default, Serialize, Deserialize, Clone, Debug)]
pub struct UserPatch {
    pub name: Option<String>,
    pub tags: Option<HashSet<TagId>>,
}

impl UserPatch {
    pub fn new() -> Self {
        Self::default()
    }
    pub fn name(mut self, v: impl Into<String>) -> Self {
        self.name = Some(v.into());
        self
    }
    pub fn tags(mut self, v: HashSet<TagId>) -> Self {
        self.tags = Some(v);
        self
    }
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub enum PlanRequest {
    RunScheduler,
    StartTask(TaskId),
    PauseTask(TaskId),
    ResumeTask(TaskId),
    CompleteTask(TaskId),
    DropTask(TaskId),
    CreateTask(Task),
    UpdateTask(TaskId, TaskPatch),
    DeleteTask(TaskId),
    CreateMilestone(Milestone),
    UpdateMilestone(MilestoneId, MilestonePatch),
    DeleteMilestone(MilestoneId),
    CreateUser(User),
    UpdateUser(UserId, UserPatch),
    DeleteUser(UserId),
    SetUserSchedule(UserId, WorkSchedule),
    ClearUserSchedule(UserId),
    SetDefaultSchedule(WorkSchedule),
    SetCalendarOverride(NaiveDate, f32),
    ClearCalendarOverride(NaiveDate),
    SetUserCalendarOverride(UserId, NaiveDate, f32),
    ClearUserCalendarOverride(UserId, NaiveDate),
    ReplacePlan(Box<Plan>),
    AddTag(String),
    RenameTag(TagId, String),
    DeleteTag(TagId),
    MoveTag(TagId, usize),
    MoveUser(UserId, usize),
    UpdatePlanSettings {
        name: String,
        start_date: NaiveDate,
        scheduler_target: NodeId,
    },
    SavePlan,
    NewPlan {
        org_id: Option<String>,
    },
    LoadPlan {
        plan_id: uuid::Uuid,
    },
    DeletePlan {
        plan_id: uuid::Uuid,
    },
    ListPlans,
    SetCurrentUser(Option<UserId>),
    ListPlanVersions {
        plan_id: uuid::Uuid,
    },
    RestorePlanVersion {
        plan_id: uuid::Uuid,
        version: String,
    },
    // Monday.com integration (handled server-side)
    MondayTestConnection {
        token: String,
        board_id: String,
    },
    MondayFetchBoardInfo {
        token: String,
        board_id: String,
    },
    MondayPull {
        plan_id: uuid::Uuid,
    },
    MondayFullReimport {
        plan_id: uuid::Uuid,
    },
    MondayPush {
        plan_id: uuid::Uuid,
    },
    /// Compute what a push would change without executing it.
    /// Returns `PlanResponse::MondayPushPreview` with op and new-item counts.
    MondayPushPreview {
        plan_id: uuid::Uuid,
    },
    SaveMondayConfig {
        plan_id: uuid::Uuid,
        config: Box<MondayConfig>,
        token: String,
    },
    LoadMondayConfig {
        plan_id: uuid::Uuid,
    },
    LoadMondayApiToken,
    // Auth user management (admin-only except ChangeMyPassword/GetUserLinks/SetUserLinks)
    GetAuthUsers,
    CreateAuthUser {
        email: String,
        password: String,
        is_admin: bool,
    },
    UpdateAuthUser {
        user_id: String,
        new_email: Option<String>,
        new_is_admin: Option<bool>,
    },
    SetAuthUserPassword {
        user_id: String,
        new_password: String,
    },
    DeleteAuthUser {
        user_id: String,
    },
    ChangeMyPassword {
        old_password: String,
        new_password: String,
    },
    // Per-plan login→plan-user links
    GetUserLinks {
        plan_id: uuid::Uuid,
    },
    SetUserLinks {
        plan_id: uuid::Uuid,
        links: Vec<UserLink>,
    },
    // Organisation management
    ListOrganisations,
    CreateOrganisation {
        name: String,
    },
    DeleteOrganisation {
        org_id: String,
    },
    RenameOrganisation {
        org_id: String,
        name: String,
    },
    GetOrgMembers {
        org_id: String,
    },
    AddOrgMember {
        org_id: String,
        user_id: String,
        role: OrgRole,
    },
    RemoveOrgMember {
        org_id: String,
        user_id: String,
    },
    SetPlanOrg {
        plan_id: uuid::Uuid,
        org_id: Option<String>,
    },
    GetPlanOrg {
        plan_id: uuid::Uuid,
    },
    // Per-plan permission management (org admins / site admins only)
    GetOrgPlans {
        org_id: String,
    },
    GetUserPlanPermissions {
        org_id: String,
        user_id: String,
    },
    SetUserPlanPermission {
        plan_id: uuid::Uuid,
        user_id: String,
        /// One of "NoAccess", "Viewer", "User", or "Default" (removes the explicit override).
        permission: String,
    },
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub enum PlanResponse {
    PlanUpdated,
    Error(PlanError),
    PlanList(Vec<(uuid::Uuid, String, String)>),
    // Monday responses
    MondayConfigLoaded(Box<MondayConfig>),
    MondayBoardInfo {
        users: Vec<MondayUser>,
        columns: Vec<BoardColumn>,
        status_labels: Vec<String>,
    },
    MondayApiToken(String),
    MondayConnected(String),
    /// Result of `MondayPushPreview`: number of field-level update ops and
    /// number of new items that would be created on Monday.
    MondayPushPreview {
        op_count: usize,
        new_item_count: usize,
    },
    // Auth responses
    AuthUsers(Vec<AuthUser>),
    UserLinks(Vec<UserLink>),
    PasswordChanged,
    AuthUserCreated {
        user_id: String,
    },
    PlanVersionList(Vec<String>),
    OrgList(Vec<Organisation>),
    OrgCreated {
        id: String,
        name: String,
    },
    OrgMembers(Vec<OrgMember>),
    PlanOrgId(Option<String>),
    OrgPlanList(Vec<(uuid::Uuid, String)>),
    UserPlanPermissions(Vec<PlanPermissionEntry>),
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub enum PlanError {
    TaskNotFound(TaskId),
    MilestoneNotFound(MilestoneId),
    UserNotFound(UserId),
    Scheduler(SchedulerError),
    Dependency(DependencyError),
    Monday(String),
    Unauthorized,
    AuthError(String),
    NoPlanActive,
}

impl std::fmt::Display for PlanError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PlanError::TaskNotFound(id) => write!(f, "task {id:?} not found"),
            PlanError::MilestoneNotFound(id) => write!(f, "milestone {id:?} not found"),
            PlanError::UserNotFound(id) => write!(f, "user {id:?} not found"),
            PlanError::Scheduler(e) => write!(f, "scheduler error: {e}"),
            PlanError::Dependency(DependencyError::Cycle) => {
                write!(f, "dependency would create a cycle")
            }
            PlanError::Dependency(DependencyError::NotFound) => {
                write!(f, "dependency target not found in plan")
            }
            PlanError::Monday(msg) => write!(f, "Monday.com error: {msg}"),
            PlanError::Unauthorized => write!(f, "Unauthorized — admin access required"),
            PlanError::AuthError(msg) => write!(f, "Auth error: {msg}"),
            PlanError::NoPlanActive => {
                write!(f, "No plan active — load or create a plan first")
            }
        }
    }
}

pub const VERSION: &str = "0.5.5";

/// Per-user server-side preferences.
#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct UserPrefs {
    pub last_plan_id: Option<uuid::Uuid>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(tag = "type")]
pub enum ServerMessage {
    Hello {
        version: String,
    },
    VersionError {
        expected: String,
        got: String,
    },
    PlanState {
        plan: Box<Plan>,
        has_monday_integration: bool,
    },
    NoPlanActive,
    Response {
        id: u64,
        response: PlanResponse,
    },
    // Monday progress — sent unsolicited while an operation is in-flight
    MondayProgress {
        done: usize,
        total: usize,
        message: String,
    },
    MondayDone {
        message: String,
    },
    MondayError {
        message: String,
    },
    // Auth
    AuthRequired,
    LoginSuccess {
        session_token: String,
        user_id: String,
        email: String,
        is_admin: bool,
        user_prefs: UserPrefs,
        org_memberships: Vec<OrgMembership>,
    },
    LoginFailed {
        message: String,
    },
}

#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(tag = "type")]
pub enum ClientMessage {
    Hello { version: String },
    Request { id: u64, request: PlanRequest },
    Login { email: String, password: String },
    Authenticate { session_token: String },
    Logout,
}

/// Apply a task patch to the plan (used for dry-run validation in the UI and
/// for actual mutations in the server engine).
pub fn apply_task_patch(plan: &mut Plan, id: TaskId, patch: TaskPatch) -> Result<(), PlanError> {
    if !plan.tasks.contains_key(&id) {
        return Err(PlanError::TaskNotFound(id));
    }
    if let Some(new_deps) = patch.dependencies {
        let old_deps = plan.tasks[&id].dependencies.clone();
        plan.tasks.get_mut(&id).unwrap().dependencies.clear();
        for dep in new_deps {
            if let Err(e) = plan.add_task_dependency(id, dep) {
                plan.tasks.get_mut(&id).unwrap().dependencies = old_deps;
                return Err(PlanError::Dependency(e));
            }
        }
    }
    // Validate new workers (if any) before mutating the task.
    if let Some(ref workers) = patch.workers {
        let task_name = plan.tasks[&id].name.clone();
        plan.validate_task_workers(&task_name, workers)
            .map_err(PlanError::Scheduler)?;
    }
    let task = plan.tasks.get_mut(&id).unwrap();
    if let Some(v) = patch.name {
        task.name = v;
    }
    if let Some(v) = patch.description {
        task.description = v;
    }
    if let Some(v) = patch.constraint {
        task.constraint = v;
    }
    if let Some(v) = patch.duration_days_target {
        task.duration_days_target = v;
    }
    if let Some(v) = patch.workers {
        task.workers = v;
    }
    if let Some(v) = patch.relaxed_mode {
        task.relaxed_mode = v;
    }
    if let Some(v) = patch.status {
        use crate::data::{TaskAllocation, TaskState};
        let ts = plan
            .node_allocations
            .tasks
            .entry(id)
            .or_insert_with(TaskState::not_started);
        ts.status = v;
        // Non-NotStarted statuses must use a Fixed allocation so the status
        // survives invalidate() which purges Dynamic (scheduler-output) entries.
        if v != crate::data::Status::NotStarted
            && matches!(ts.allocation, TaskAllocation::Dynamic { .. })
        {
            let sentinel = chrono::NaiveDate::from_ymd_opt(1970, 1, 1).unwrap();
            let start = plan
                .tasks
                .get(&id)
                .and_then(|t| t.actual_start)
                .unwrap_or(sentinel);
            ts.allocation = TaskAllocation::Fixed {
                start_date: start,
                end_date: start,
                corrected_end_date: None,
                time_allocation: vec![],
            };
        }
    }
    if let Some(v) = patch.actual_start_date {
        plan.set_task_actual_start(id, v);
    }
    if let Some(v) = patch.actual_end_date {
        plan.set_task_actual_end(id, v);
    }
    plan.node_allocations.invalidate();
    plan.simplify_all_dependencies();
    Ok(())
}

/// Apply a milestone patch to the plan.
pub fn apply_milestone_patch(
    plan: &mut Plan,
    id: MilestoneId,
    patch: MilestonePatch,
) -> Result<(), PlanError> {
    if !plan.milestones.contains_key(&id) {
        return Err(PlanError::MilestoneNotFound(id));
    }
    if let Some(new_deps) = patch.dependencies {
        let old_deps = plan.milestones[&id].dependencies.clone();
        plan.milestones.get_mut(&id).unwrap().dependencies.clear();
        for dep in new_deps {
            if let Err(e) = plan.add_milestone_dependency(id, dep) {
                plan.milestones.get_mut(&id).unwrap().dependencies = old_deps;
                return Err(PlanError::Dependency(e));
            }
        }
    }
    let ms = plan.milestones.get_mut(&id).unwrap();
    if let Some(v) = patch.name {
        ms.name = v;
    }
    if let Some(v) = patch.description {
        ms.description = v;
    }
    if let Some(v) = patch.constraint {
        ms.constraint = v;
    }
    plan.node_allocations.invalidate();
    plan.simplify_all_dependencies();
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn clearing_constraint_via_json_roundtrip() {
        // Some(None) must survive JSON serialisation so the server can
        // distinguish "clear constraint" from "no change".
        let patch = MilestonePatch::new().constraint(None);
        let json = serde_json::to_string(&patch).unwrap();
        let decoded: MilestonePatch = serde_json::from_str(&json).unwrap();
        assert!(
            decoded.constraint == Some(None),
            "constraint should round-trip as Some(None) (clear), got {:?}",
            decoded.constraint
        );

        let patch2 = TaskPatch::new().constraint(None);
        let json2 = serde_json::to_string(&patch2).unwrap();
        let decoded2: TaskPatch = serde_json::from_str(&json2).unwrap();
        assert!(
            decoded2.constraint == Some(None),
            "constraint should round-trip as Some(None) (clear), got {:?}",
            decoded2.constraint
        );
    }
}

File: plinko-web/src/protocol.ts
/**
 * TypeScript mirror of `plinko-shared/src/protocol.rs` and associated data types.
 *
 * Serialisation rules:
 *  - Rust tuple-newtype IDs (e.g. `TaskId(Uuid)`) serialize as plain UUID strings.
 *  - Rust enums with `#[serde(tag = "type")]` become discriminated unions keyed on `type`.
 *  - Rust enums *without* a tag attribute use the default serde encoding:
 *      unit variants  → bare string  (e.g. `"PlanStart"`)
 *      struct/tuple variants → `{ "VariantName": { ...fields } }`
 *  - `HashMap<K, V>` with string-serialisable keys → `Record<string, V>`.
 *  - `NaiveDate` → ISO-8601 date string (`"YYYY-MM-DD"`).
 *  - `Option<T>` → `T | null` in JSON; `undefined` means the field is absent.
 */

// ── Primitive aliases ────────────────────────────────────────────────────────

export type TaskId = string;       // UUID
export type MilestoneId = string;  // UUID
export type UserId = string;       // UUID
export type TagId = string;        // UUID
export type IsoDate = string;      // "YYYY-MM-DD"

// ── NodeId ───────────────────────────────────────────────────────────────────
// Serde default (no tag): unit variant is bare string, struct variants are wrapped.
export type NodeId =
  | "PlanStart"
  | { Task: TaskId }
  | { Milestone: MilestoneId };

// ── Constraint ───────────────────────────────────────────────────────────────

export type ConstraintKind = "Fixed" | "Earliest" | "Latest";

export interface DateConstraint {
  date: IsoDate;
  kind: ConstraintKind;
}

// ── Dependency ───────────────────────────────────────────────────────────────

export interface Dependency {
  id: NodeId;
  lag_days: number;
}

// ── Worker slots ─────────────────────────────────────────────────────────────
// Serde default enum encoding: { "Specific": { user_id, workload_days } }

export type WorkerSlot =
  | { Specific: { user_id: UserId; workload_days: number } }
  | { Placeholder: { required_tags: TagId[]; workload_days: number } };

// ── Status ───────────────────────────────────────────────────────────────────

export type Status =
  | "NotStarted"
  | "InProgress"
  | "OnHold"
  | "Complete"
  | "Dropped";

// ── Task ─────────────────────────────────────────────────────────────────────

export interface Task {
  id: TaskId;
  name: string;
  description: string;
  dependencies: Dependency[];
  workers: WorkerSlot[];
  constraint: DateConstraint | null;
  duration_days_target: number;
  relaxed_mode: boolean;
  actual_start: IsoDate | null;
  context_label: string | null;
}

// ── Milestone ────────────────────────────────────────────────────────────────

export interface Milestone {
  id: MilestoneId;
  name: string;
  description: string;
  dependencies: Dependency[];
  constraint: DateConstraint | null;
  context_label: string | null;
}

// ── Tag ──────────────────────────────────────────────────────────────────────

export interface Tag {
  id: TagId;
  name: string;
}

// ── User ─────────────────────────────────────────────────────────────────────

export interface User {
  id: UserId;
  name: string;
  tags: TagId[];
}

// ── WorkSchedule ─────────────────────────────────────────────────────────────
// Weekday enum serialises as string.

export type Weekday =
  | "Monday"
  | "Tuesday"
  | "Wednesday"
  | "Thursday"
  | "Friday"
  | "Saturday"
  | "Sunday";

export interface WorkSchedule {
  days: Partial<Record<Weekday, number>>;
}

// ── UserData ─────────────────────────────────────────────────────────────────

export interface UserData {
  user: User;
  schedule: WorkSchedule | null;
}

// ── CalendarOverrides ────────────────────────────────────────────────────────

export interface CalendarOverrides {
  entries: Record<IsoDate, number>;
}

// ── Allocation types ─────────────────────────────────────────────────────────

export interface WorkSegment {
  user: UserId;
  date: IsoDate;
  hours_worked: number;
}

export type TaskAllocation =
  | {
      Fixed: {
        start_date: IsoDate;
        end_date: IsoDate;
        corrected_end_date: IsoDate | null;
        time_allocation: WorkSegment[];
      };
    }
  | {
      Dynamic: {
        scheduled_start_date: IsoDate;
        scheduled_end_date: IsoDate;
        time_allocation: WorkSegment[];
      };
    };

export interface TaskState {
  status: Status;
  allocation: TaskAllocation;
}

export interface MilestoneAllocation {
  date: IsoDate;
  derived_status: Status;
}

export interface ConstraintViolation {
  node_name: string;
  kind: ConstraintKind;
  required_date: IsoDate;
  scheduled_date: IsoDate;
}

// NodeAllocations: constraint_violations uses NodeId-string keys
export interface NodeAllocations {
  tasks: Record<TaskId, TaskState>;
  milestones: Record<MilestoneId, MilestoneAllocation>;
  constraint_violations: Record<string, ConstraintViolation>;
}

// ── Plan ─────────────────────────────────────────────────────────────────────

export interface Plan {
  id: string; // Uuid
  name: string;
  users_data: Record<UserId, UserData>;
  user_order: UserId[];
  tags: Tag[];
  tasks: Record<TaskId, Task>;
  milestones: Record<MilestoneId, Milestone>;
  start_date: IsoDate;
  default_schedule: WorkSchedule;
  calendar: CalendarOverrides;
  user_calendar_overrides: Record<UserId, CalendarOverrides>;
  scheduler_target: NodeId;
  node_allocations: NodeAllocations;
}

// ── Monday config types ───────────────────────────────────────────────────────

export interface ColumnMap {
  person_column_id: string;
  status_column_id: string;
  dependency_column_id: string;
  workload_column_id: string;
  timeline_column_id: string;
}

export interface UserMapping {
  monday_user_id: string;
  monday_name: string;
  plinko_user_id: UserId | null;
}

export interface StatusMapping {
  monday_label: string;
  plinko_status: Status;
}

export interface ItemNodeMapping {
  monday_item_id: string;
  plinko_node_id: NodeId;
  board_id: string;
}

export interface MondayConfig {
  board_id: string;
  column_map: ColumnMap;
  user_mappings: UserMapping[];
  status_mappings: StatusMapping[];
  item_node_map: ItemNodeMapping[];
  use_subitems: boolean;
  workload_in_hours: boolean;
  show_monday_context: boolean;
}

export interface BoardColumn {
  id: string;
  title: string;
  column_type: string;
}

export interface MondayUser {
  id: string;
  name: string;
  email: string;
}

// ── Protocol: patches ────────────────────────────────────────────────────────

export interface TaskPatch {
  name?: string;
  description?: string;
  status?: Status;
  actual_start_date?: IsoDate | null;
  actual_end_date?: IsoDate | null;
  constraint?: DateConstraint | null;
  duration_days_target?: number;
  workers?: WorkerSlot[];
  dependencies?: Dependency[];
  relaxed_mode?: boolean;
}

export interface MilestonePatch {
  name?: string;
  description?: string;
  constraint?: DateConstraint | null;
  dependencies?: Dependency[];
}

export interface UserPatch {
  name?: string;
  tags?: TagId[];
}

// ── Auth types ────────────────────────────────────────────────────────────────

export interface AuthUser {
  id: string;
  email: string;
  is_admin: boolean;
}

export interface UserLink {
  login_user_id: string;
  plan_user_id: UserId;
}

// ── Protocol: PlanRequest ─────────────────────────────────────────────────────
// Serde default enum encoding. Unit variants are bare strings;
// tuple/struct variants are `{ "VariantName": payload }`.

export type PlanRequest =
  | "RunScheduler"
  | "SavePlan"
  | "NewPlan"
  | "ListPlans"
  | { StartTask: TaskId }
  | { PauseTask: TaskId }
  | { ResumeTask: TaskId }
  | { CompleteTask: TaskId }
  | { DropTask: TaskId }
  | { CreateTask: Task }
  | { UpdateTask: [TaskId, TaskPatch] }
  | { DeleteTask: TaskId }
  | { CreateMilestone: Milestone }
  | { UpdateMilestone: [MilestoneId, MilestonePatch] }
  | { DeleteMilestone: MilestoneId }
  | { CreateUser: User }
  | { UpdateUser: [UserId, UserPatch] }
  | { DeleteUser: UserId }
  | { SetUserSchedule: [UserId, WorkSchedule] }
  | { ClearUserSchedule: UserId }
  | { SetDefaultSchedule: WorkSchedule }
  | { SetCalendarOverride: [IsoDate, number] }
  | { ClearCalendarOverride: IsoDate }
  | { SetUserCalendarOverride: [UserId, IsoDate, number] }
  | { ClearUserCalendarOverride: [UserId, IsoDate] }
  | { ReplacePlan: Plan }
  | { AddTag: string }
  | { RenameTag: [TagId, string] }
  | { DeleteTag: TagId }
  | { MoveTag: [TagId, number] }
  | { MoveUser: [UserId, number] }
  | { UpdatePlanSettings: { name: string; start_date: IsoDate; scheduler_target: NodeId } }
  | { LoadPlan: { plan_id: string } }
  | { DeletePlan: { plan_id: string } }
  | { SetCurrentUser: UserId | null }
  | { MondayTestConnection: { token: string; board_id: string } }
  | { MondayFetchBoardInfo: { token: string; board_id: string } }
  | { MondayPull: { plan_id: string } }
  | { MondayFullReimport: { plan_id: string } }
  | { MondayPush: { plan_id: string } }
  | { SaveMondayConfig: { plan_id: string; config: MondayConfig; token: string } }
  | { LoadMondayConfig: { plan_id: string } }
  | "LoadMondayApiToken"
  // Auth
  | "GetAuthUsers"
  | { CreateAuthUser: { email: string; password: string; is_admin: boolean } }
  | { UpdateAuthUser: { user_id: string; new_email?: string; new_is_admin?: boolean } }
  | { SetAuthUserPassword: { user_id: string; new_password: string } }
  | { DeleteAuthUser: { user_id: string } }
  | { ChangeMyPassword: { old_password: string; new_password: string } }
  | { GetUserLinks: { plan_id: string } }
  | { SetUserLinks: { plan_id: string; links: UserLink[] } }
  | { GetPlanVisibility: { plan_id: string } }
  | { SetPlanVisibility: { plan_id: string; user_ids: string[] } }
  | { ListPlanVersions: { plan_id: string } }
  | { RestorePlanVersion: { plan_id: string; version: string } };

// ── Protocol: PlanResponse ────────────────────────────────────────────────────

export type PlanResponse =
  | "PlanUpdated"
  | "PasswordChanged"
  | { Error: PlanError }
  | { PlanList: [string, string, string][] }
  | { MondayConfigLoaded: MondayConfig }
  | { MondayBoardInfo: { users: MondayUser[]; columns: BoardColumn[]; status_labels: string[] } }
  | { MondayApiToken: string }
  | { MondayConnected: string }
  | { AuthUsers: AuthUser[] }
  | { UserLinks: UserLink[] }
  | { AuthUserCreated: { user_id: string } }
  | { PlanVisibility: { plan_id: string; user_ids: string[] } }
  | { PlanVersionList: string[] };

export type SchedulerError =
  | "EmptyChain"
  | { MissingTaskAffinity: { task_name: string; required_tags: string[] } }
  | { SpecificWorkerNotFound: { task_name: string; user_id: string } }
  | { NoPathsToNode: unknown }
  | { DisconnectedNode: unknown };

export type PlanError =
  | { TaskNotFound: TaskId }
  | { MilestoneNotFound: MilestoneId }
  | { UserNotFound: UserId }
  | { Scheduler: SchedulerError }
  | { Dependency: "Cycle" | "NotFound" }
  | { Monday: string }
  | "Unauthorized"
  | { AuthError: string }
  | "NoPlanActive";

export function formatPlanError(err: PlanError): string {
  if (typeof err === "string") return err; // "Unauthorized", "NoPlanActive"
  if ("TaskNotFound" in err) return "Task not found.";
  if ("MilestoneNotFound" in err) return "Milestone not found.";
  if ("UserNotFound" in err) return "User not found.";
  if ("Dependency" in err)
    return err.Dependency === "Cycle"
      ? "Adding this dependency would create a cycle."
      : "Dependency not found.";
  if ("Monday" in err) return `Monday.com error: ${err.Monday}`;
  if ("AuthError" in err) return err.AuthError;
  if ("Scheduler" in err) {
    const se = err.Scheduler;
    if (se === "EmptyChain") return "Scheduler error: empty node chain.";
    if (typeof se === "object" && "MissingTaskAffinity" in se) {
      const { task_name, required_tags } = se.MissingTaskAffinity;
      if (required_tags.length === 0)
        return `Cannot create task "${task_name}": no users exist in this plan. Add a user first.`;
      return `Cannot create task "${task_name}": no user has all required tags (${required_tags.join(", ")}).`;
    }
    if (typeof se === "object" && "SpecificWorkerNotFound" in se) {
      return `Cannot create task "${se.SpecificWorkerNotFound.task_name}": a selected worker is not in this plan.`;
    }
    if (typeof se === "object" && "DisconnectedNode" in se) {
      return "One or more dependencies reference a node that doesn't exist in the plan.";
    }
    return `Scheduler error: ${JSON.stringify(se)}`;
  }
  return JSON.stringify(err);
}

export interface UserPrefs {
  last_plan_id: string | null;
}

// ── Protocol: ServerMessage (`#[serde(tag = "type")]`) ───────────────────────

export type ServerMessage =
  | { type: "Hello"; version: string }
  | { type: "VersionError"; expected: string; got: string }
  | { type: "PlanState"; plan: Plan; has_monday_integration: boolean }
  | { type: "NoPlanActive" }
  | { type: "Response"; id: number; response: PlanResponse }
  | { type: "MondayProgress"; done: number; total: number; message: string }
  | { type: "MondayDone"; message: string }
  | { type: "MondayError"; message: string }
  | { type: "AuthRequired" }
  | { type: "LoginSuccess"; session_token: string; user_id: string; email: string; is_admin: boolean; user_prefs: UserPrefs }
  | { type: "LoginFailed"; message: string };

// ── Protocol: ClientMessage (`#[serde(tag = "type")]`) ───────────────────────

export type ClientMessage =
  | { type: "Hello"; version: string }
  | { type: "Request"; id: number; request: PlanRequest }
  | { type: "Login"; email: string; password: string }
  | { type: "Authenticate"; session_token: string }
  | { type: "Logout" };

// ── Protocol version ─────────────────────────────────────────────────────────

export const PROTOCOL_VERSION = "0.3.4";
